src/Language/PureScript/TypeChecker/Entailment.hs

```diff
@@ -406,56 +406,65 @@ entails SolverOptions{..} constraint context hints =
+      , i > 0 && j > 0 = do
+          -- If both arguments have kind @k1 -> k2@ (e.g. @data D a b = D a@
+          -- in the constraint @Coercible (D a) (D a')@), yield a new wanted
+          -- constraint in terms of the types saturated with the same variables
+          -- (e.g. @Coercible (D a t0) (D a' t0)@ in the exemple).
+          let aTyVars = S.fromList $ usedTypeVariables a
+              bTyVars = S.fromList $ usedTypeVariables b
+              saturate' = saturate $ aTyVars `S.union` bTyVars
+          pure [srcCoercibleConstraint (saturate' a i) (saturate' b j)]
+      | (TypeConstructor _ aTyName, _, axs) <- unapplyTypes a
+      , (TypeConstructor _ bTyName, _, bxs) <- unapplyTypes b
+      , not (null axs) && not (null bxs) && aTyName == bTyName = do
+          -- If both arguments are applications of the same type constructor
+          -- (e.g. @data D a b = D a@ in the constraint
+          -- @Coercible (D a b) (D a' b')@), infer the roles of the type
+          -- constructor's arguments and generate wanted constraints
+          -- appropriately (e.g. here @a@ is representational and @b@ is
+          -- phantom, yielding @Coercible a a'@).
+          let tyRoles = lookupEnvRoles env aTyName
+              k role ax bx = case role of
+                Nominal
+                  -- If we had first-class equality constraints, we'd just
+                  -- emit one of the form @(a ~ b)@ here and let the solver
+                  -- recurse. Since we don't we must compare the types at
+                  -- this point and fail if they don't match. This likely
+                  -- means there are cases we should be able to handle that
+                  -- we currently can't, but is at least sound.
+                  | ax == bx ->
+                      Just []
+                  | otherwise ->
+                      Nothing
+                Representational ->
+                  Just [srcCoercibleConstraint ax bx]
+                Phantom ->
+                  Just []
+          fmap concat $ sequence $ zipWith3 k tyRoles axs bxs
+      | (TypeConstructor _ tyName, _, xs) <- unapplyTypes a
+      , Just (tvs, wrappedTy, _) <- lookupNewtypeConstructor env tyName = do
+          -- If the first argument is a newtype applied to some other types
+          -- (e.g. @newtype T a = T a@ in @Coercible (T X) b@), look up the
+          -- type of its wrapped field and yield a new wanted constraint in
+          -- terms of that type with the type arguments substituted in (e.g.
+          -- @Coercible (T[X/a]) b = Coercible X b@ in the example).
+          let wrappedTySub = replaceAllTypeVars (zip tvs xs) wrappedTy
+          pure [srcCoercibleConstraint wrappedTySub b]
+      | otherwise =
         -- In all other cases we can't solve the constraint.
         Nothing
 
+    srcCoercibleConstraint :: SourceType -> SourceType -> SourceConstraint
+    srcCoercibleConstraint a b = srcConstraint C.Coercible [] [a, b] Nothing
```

Comment:
If Coercible is PolyKinded, the kind applications need to be supplied. Currently you are always setting this to an empty `[]` which is problematic.

Reply:
I’ve added a kind argument to `srcCoercibleConstraint` and the type class dictionaries returned by `solveCoercible` now hold the kinds list `forClassName` is called with. Can this affect the solving of `Coercible` instances in the absence of syntax for kind application?

Reply:
In general when solving constraints, we need to unify kinds in substitutions https://github.com/purescript/purescript/blob/5e3f01d892bcf23a7c3775274e6fb8ba5a214851/src/Language/PureScript/TypeChecker/Entailment.hs#L288-L319

It is not _likely_ that solving a `Coercible` constraint and ignoring the kinds will cause problems in 99.9% of cases. I would not be surprised though if you can construct something that relies on kind unification in order to typecheck, however unlikely. Unification can be very non-local. I think that we should make an effort to construct kinds/type-correct terms though. It can only go wrong if you don't 😆

Reply:
I tried to reject heterogeneously kinded `Coercible` constraints by unifying the arguments kinds with:

```hs
(_, kind) <- kindOf a
(_, kind') <- kindOf b
unifyKinds kind kind'
```

but a few tests then throw `UndefinedTypeVariable` errors.

I guess I should compute and apply a substitution like it’s done in `unifySubstKind` but I’m kind of lost here :grimacing:

Reply:
This happened when inferring the kinds of higher kinded types saturated in `coercibleWanteds` because I wasn’t adding the generated type variables to the environment. I went with `freshTypeWithKind` instead to reuse the existing machinery.